### PR TITLE
Adding optional ES2015 support via Babel

### PR DESCRIPTION
**Pull request by nicolasgarnier**
Currently the way to opt-in is to rename a new gulp task. let me know if you think of better ways.

We could also simply enable it by default since this workflow also works with ES<=5 files.

@addyosmani PTAL

### CODE REVIEW

**Comment by addyosmani** [comment]
Fwiw, @gauntface has been exploring ES6 support via Babel as part of the canonical apps project and has some ideas he wanted to share on different ways we could tackle this. Approaches off the top of my head:

* Make Babel the default. Our build pipeline works fine for both ES5 and ES6 source without the end-user needing to do anything. In this option, we would also avoid using an `.es6` extension which some folks use. I personally would love this to be the default way of tackling ES6 support.
* Make Babel opt-in. Similar to this PR. We decide on an approach and make it a simple case of commenting/uncommenting.
* Make Babel extension specific. Support it by default, but only pick up files with an `.`es6` or similar extension so you're separating your ES6 and ES5 build pipelines.

Looping in @thejameskyle as well as he might have an interest in this.

**Comment by gauntface** [comment]
LGTM from me.

I'd vote just making this the default and only script task.

It's a shame that this, like the rest of WSK, requires hard coding file paths and names to work.

**Comment by nicolasgarnier** [comment]
I just updated the `script` gulp task, comments and readme as per @addyosmani's first comment.

I'd also vote for making this the only and default gulp task.

I agree it would also be nice not to have to hardcode the name of the scripts in the gulpfile. I could probably investigate to see how we could solve this using useref on a separate issue.

File: gulpfile.babel.js [hunk 1]
```diff
@@ -125,6 +128,31 @@ gulp.task('scripts', () =>
...
+//      // Note: Since we are not using useref in the scripts build pipeline,
+//      //       you need to explicitly list your scripts here in the right order
+//      //       to be correctly concatenated
+//      './app/scripts/main.js',
+//      './app/scripts/person.es6.js'
+//      // Other scripts
+//    ])
+//      .pipe($.sourcemaps.init())
+//      .pipe($.changed('.tmp/scripts', {extension: '.js'}))
+//      .pipe(babel({retainLines: true}))
```
Comment (thejameskyle) [inline gulpfile.babel.js]: It's better to do this config in a `.babelrc` file.

Comment (thejameskyle) [inline gulpfile.babel.js]: Also any reason you aren't using `$.babel` like the rest of the gulpfile?

**Comment by thejameskyle** [comment]
I certainly think it would be great by default (I may be a tad bit biased) as there's really no harm if people want to continue only using <=ES5 features.

File: gulpfile.babel.js (hunk 1 above)
Comment (nicolasgarnier) [inline gulpfile.babel.js]: Changed to using `$.babel` and `.babelrc` thank!

**Comment by nicolasgarnier** [comment]
The `.babelrc` file support gave me a much better/cleaner (I think) way to opt-in ES6 support. Now:

> To enable ES6 support remove the line `"only": "gulpfile.babel.js",` in the [.babelrc](.babelrc) file.

I'd still prefer having this enabled by default though ;)
The only down-side is that Babel does modify ES5 files a bit. This doesn't make a difference if you use a browser that supports sourcemaps though...

File: .jscsrc [hunk 2]
```diff
@@ -1,8 +1,6 @@
 {
   "preset": "google",
   "disallowSpacesInAnonymousFunctionExpression": null,
-  "excludeFiles": [
-    "node_modules/**",
-    "gulpfile.babel.js"
-  ]
+  "esnext": true,
+  "excludeFiles": ["node_modules/**"]
```
Comment (hzoo) [inline .jscsrc]: jscs excludes `node_modules` by default now (although it looks like we need to update the docs on this: http://jscs.info/overview.html#excludefiles, which I'l do now)

File: README.md [hunk 3]
```diff
@@ -56,6 +56,12 @@ If you find yourself running into issues during installation or running the tool
 
 If you would prefer not to use any of our tooling, delete the following files from the project: `package.json`, `gulpfile.babel.js`, `.jshintrc` and `.travis.yml`. You can now safely use the boilerplate with an alternative build-system or no build-system at all if you choose.
 
+## ES6 Support
+
+Web Starter Kit has optional ES6 support using [Babel](https://babeljs.io/).
+To enable ES6 support remove the line `"only": "gulpfile.babel.js",` in the [.babelrc](.babelrc) file.
+ES6 source code will be automatically transpiled to ES5 for wide browser support.
+
```
Comment (sindresorhus) [inline README.md]: ES6 => ES2015

ES2015 is the official name now.

Comment (nicolasgarnier) [inline README.md]: Changed! thx

File: .jscsrc (hunk 2 above)
Comment (addyosmani) [inline .jscsrc]: @nicolasgarnier can we update to drop the excludeFiles line here as `node_modules` are excluded by default per @hzoo?

Comment (nicolasgarnier) [inline .jscsrc]: Done!

**Comment by addyosmani** [comment]
Can we rebase this against master please? :tophat:

**Comment by nicolasgarnier** [comment]
Done!

**Comment by addyosmani** [comment]
Can we get the commits here squashed? :steam_locomotive:

**Comment by nicolasgarnier** [comment]
Done!

**Comment by timmywil** [comment]
Could this be landed?

File: .babelrc [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
+  // Remove the line below to enable ES6 support.
```
Comment (arthurvr) [inline .babelrc]: `ES2015`

Comment (nicolasgarnier) [inline .babelrc]: Done

**Pull request merged**